### ISSUE TITLE
Editing

### ISSUE THREAD

**Issue by sbrunner** [camptocamp/c2cgeoportal#1662]
Spec: https://github.com/camptocamp/c2cgeoportal/wiki/Spec-%231662-Editing

**Comment by sbrunner** [comment]
Requires: https://github.com/camptocamp/c2cgeoportal/issues/2040

**Comment by ybolognini** [comment]
## Feedbacks via mail

From @gnerred
J'ai juste un point négatif à mentionner, qu'il faudrait corriger dans la 2 : lorsqu'on a un attribut de type date et qu'on clique dans le champs sans n'y entrer aucune valeur, le module considère qu'un string vide a été entré, et cela fait planter l'enregistrement. C'est le seul problème qui m'a été remonté par les utilisateurs, mais ça peut être relativement handicapant. Surtout pour les utilisateurs qui utilisent le bouton TAB pour passer d'un attribut à un autre.

From @vvmalbre
Unsere Frage ist, ob und wie das Editing-Tool in einem Mehrserverbetrieb eingesetzt werden kann.

From @svavoi

### Questions to the specification:
* Does the drawing of the geometry still work the same? For example: double-click to finish a line or a polygon, ESC to cancel the drawing
* Can attributes still be hidden from the right panel with “exclude_properties”?

### Improvements for current editing interface:
* It would be nice to see what kind of attribute is expected in the attribute fields. At the moment you need to click into the field to see if it is a dropdown, a text or a Boolean field. Why not show it straight away?
* When you click into a field or change an attribute you get a red triangle in the upper left corner of the field but it is not evident what this means. A note what it means would be nice on hoover. Or a symbol right to the field with green if there is no change of the attribute, yellow if the attribute has been changed and red if the attribute is missing and mandatory or if there is an error with this field. Or with symbols with a tick, an exclamation mark or a cross.
* It would be nice to see which fields are mandatory to fill in. At the moment you only get an error message when you try to save the geometry.
* When there is a constraint on an attribute and you get an error because you filled in a wrong value, the value should be marked which provoked it. And show what the constraint is (if possible).
* Possibility to search in the dropdown list by entering a value/ a letter in the field
* At the moment the user can’t see what snapping layers are configured. It would be nice if the user can see what is available and then select the one he wants to use.
* Deleting a vertex is not that intuitive with clicking and pressing DELETE. Would it be better to make a right click and get the additional action menu where you can delete the vertex?

### Further improvements for future editing interface:
* It would be nice to be able to set attributes read-only for the right panel. For example for attributes that are calculated by a trigger. This values should be greyed out.
* Drawing of multipolygons/multipoints/multiline, could be an additional action like “cut geometry”
* Drawing of an arc or a circle
* Possibility to show area size when you draw a feature
* Possibility to draw a line or a polygon of a specific size
* Possibility to draw rectangular angles (or user defined angles)
* Possibility to set a point to a specific coordinate
* Copy attributes form another geometry
* Copy a geometry from a different (not editable) layer. Maybe they could be configured like snapping layers?
* We have editing layers that hold a lot of attributes (>10 attributes), it would be good to have a possibility to open another window with an additional form (for example an external plugin) to set the non-geometry attributes.

**Comment by ybolognini** [comment]
@gnerred date problem

This will be taken into account.

@vvmalbre multi servers environment

Editing, like short-URLs, needs to have a single database or a way to synchronize multiple databases. At Camptocamp we are working on such solutions, but it is more an infrastructure problem than a GeoMapFish problem. 

@svavoi Does the drawing of the geometry still work the same? For example: double-click to finish a line or a polygon, ESC to cancel the drawing

Behavior is the same as for redlining: double-click finishes a line or a polygon, but for now Esc doesn't cancel.

@svavoi Can attributes still be hidden from the right panel with “exclude_properties”?

Yes. This is a server configuration, so it will be automatically included in 2.1.

@svavoi It would be nice to see what kind of attribute is expected in the attribute fields. At the moment you need to click into the field to see if it is a dropdown, a text or a Boolean field. Why not show it straight away?

We'll try to include this in the new version. ping @pgiraud

@svavoi When you click into a field or change an attribute you get a red triangle in the upper left corner of the field but it is not evident what this means

We will make a proposition. ping @pgiraud 

@svavoi It would be nice to see which fields are mandatory to fill in. At the moment you only get an error message when you try to save the geometry
@svavoi When there is a constraint on an attribute and you get an error because you filled in a wrong value, the value should be marked which provoked it. And show what the constraint is (if possible)

These errors come directly from the database. Now we are not able to have this information (what field provoked the error) on the client. But we could think about a solution as an enhancement.

@svavoi Possibility to search in the dropdown list by entering a value/ a letter in the field

This will be included only if the dropdown widget we use allows it. If this isn't the case, this could be an enhancement.

@svavoi At the moment the user can’t see what snapping layers are configured. It would be nice if the user can see what is available and then select the one he wants to use

We'll see what's possible, but it will stay simple. ping @pgiraud 

@svavoi Deleting a vertex is not that intuitive with clicking and pressing DELETE. Would it be better to make a right click and get the additional action menu where you can delete the vertex?

We use standard OpenLayers 3 for delete a vertex (simple click). But yes, an action menu would be nice. I'll add this as an enhancement.

I'll create a new issue with the "Further improvements" list.

**Comment by ybolognini** [comment]
@svavoi It would be nice to see what kind of attribute is expected in the attribute fields. At the moment you need to click into the field to see if it is a dropdown, a text or a Boolean field. Why not show it straight away?

We will do something to go in this direction. It also depends on what is technically possible with the tools we use. 

@svavoi When you click into a field or change an attribute you get a red triangle in the upper left corner of the field but it is not evident what this means

This triangle was an ExtJS feature. What is planned for the new version is that the Save button will be active only when field values have been updated.

@svavoi At the moment the user can’t see what snapping layers are configured. It would be nice if the user can see what is available and then select the one he wants to use

In version 1.6, snapping configuration is quite simple: integrator can decide what layers are used for snapping. Displaying and choosing snapping layers is a possible enhancement. I'll add this to the list.

@rbovard @gnerred @svarim @vvmalbre @mriedo @gsollberger @svavoi @cedricmoullet @kalbermattenm

To all: Now ready for validation

**Comment by svavoi** [comment]
Thanks @ybolognini this is ok for me.

**Comment by ybolognini** [comment]
@rbovard @gnerred @svarim @vvmalbre @mriedo @gsollberger @svavoi @cedricmoullet @kalbermattenm

Last call! :)

**Comment by rbovard** [comment]
Blank vote for me.

**Comment by gnerred** [comment]
ok for me

**Comment by sbrunner** [comment]
Fixed with #1922
